### PR TITLE
Move function topic value from labels to annotations

### LINKED ISSUE #3: Move to annotations
OpenFaaS now has support for annotations meaning we can use a rich value for the topic list.

- faas-swarm 0.4.0
- faas-netes 0.6.0
- openfaas-operator (awaiting review & merge)

This task means changing the code to look at the function's topic value in an annotation instead of a label.

It should be a straight-forward change plus basic smoke testing on Kubernetes and Swarm.

### PR DESCRIPTION
**Pull request by ivanayov**
This change comes to addopt OpenFaaS support for annotations
meaning that a rich value for the topic list can be used

Signed-off-by: Ivana Yovcheva <iyovcheva@vmware.com>

Resolves #3 

### How has this been tested?

`stack.yml`:
```
provider:
  name: faas
  gateway: http://127.0.0.1:8080

functions:
  kafka-test:
    lang: python
    handler: ./kafka-test
    image: kafka-test
    annotations:
      topic: faas-request
    basic_auth: false
```

```
$ ./build.sh
$ SERVICE_NAME=kafka_kafka
$ TASK_ID=$(docker service ps --filter 'desired-state=running' $SERVICE_NAME -q)
$ CONTAINER_ID=$(docker inspect --format '{{ .Status.ContainerStatus.ContainerID }}' $TASK_ID)
$ docker exec -it $CONTAINER_ID kafka-console-producer --broker-list kafka:9092 --topic faas-request

Some random request here

$ docker service logs kafka_connector -f

kafka_connector.1.4w6hi1loqsa2@linuxkit-025000000001    | 2018/08/06 15:30:19 Binding to topics: [faas-request]
kafka_connector.1.4w6hi1loqsa2@linuxkit-025000000001    | Rebalanced: &{Type:rebalance start Claimed:map[] Released:map[] Current:map[]}
kafka_connector.1.4w6hi1loqsa2@linuxkit-025000000001    | Rebalanced: &{Type:rebalance OK Claimed:map[faas-request:[0]] Released:map[] Current:map[faas-request:[0]]}
kafka_connector.1.4w6hi1loqsa2@linuxkit-025000000001    | 2018/08/06 15:30:22 Syncing topic map

kafka_connector.1.4w6hi1loqsa2@linuxkit-025000000001    | [#2] Received on [faas-request,0]: 'Some random request here'
kafka_connector.1.4w6hi1loqsa2@linuxkit-025000000001    | 2018/08/06 15:32:09 Invoke function: kafka-test
kafka_connector.1.4w6hi1loqsa2@linuxkit-025000000001    | 2018/08/06 15:32:09 Response [200] from kafka-test

```

### CODE REVIEW

**Comment by alexellis** [comment]
Can you show how you configured the function and that it got invoked?

**Comment by ivanayov** [comment]
Thanks for noticing @alexellis 
I thought that I've wrote it, but seems that I've forgot

**Comment by alexellis** [comment]
I think you'll need to vendor with prune. Many unrelated files have been added in your PR

**Comment by alexellis** [comment]
Best to reset and push again rather than adding additional commits

**Comment by ivanayov** [comment]
@alexellis It's addressed now

File: types/function_list_builder_test.go [hunk 1]
```diff
@@ -17,12 +17,12 @@ func TestBuildSingleMatchingFunction(t *testing.T) {
 	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 
 		functions := []requests.Function{}
-		labelMap := make(map[string]string)
-		labelMap["topic"] = "topic1"
+		annotationlMap := make(map[string]string)
+		annotationlMap[AnnotationLabelPrefix+"topic"] = "topic1"
```
Comment (alexellis) [inline types/function_list_builder_test.go]: I don't think this value is present in the annotations.. I would expect it to be hidden and internal. cc @ewilde

File: types/function_list_builder.go [hunk 2]
```diff
@@ -45,10 +47,10 @@ func (s *FunctionLookupBuilder) Build() (map[string][]string, error) {
 	}
 
 	for _, function := range functions {
-		if *function.Labels != nil {
-			labels := *function.Labels
+		if function.Annotations != nil {
+			annotations := *function.Annotations
 
-			if topic, pass := labels["topic"]; pass {
+			if topic, pass := annotations[AnnotationLabelPrefix+"topic"]; pass {
```
Comment (alexellis) [inline types/function_list_builder.go]: I don't think this is right... the prefix is not required, it's an internal and hidden label on Swarm only.

Comment (ewilde) [reply types/function_list_builder.go]: Correct prefix is not needed

File: types/function_list_builder_test.go (hunk 1 above)
Comment (ewilde) [reply types/function_list_builder_test.go]: @alexellis correct, you don't need to add this prefix. Depending on the configured provider annotations get stored differently. In swarm we would store this as a label and add `com.openfaas.annotations.` on your behalf internally and remove it when queried

So in short prefix not needed 😄

File: types/function_list_builder_test.go [hunk 3]
```diff
@@ -17,12 +17,12 @@ func TestBuildSingleMatchingFunction(t *testing.T) {
 	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 
 		functions := []requests.Function{}
-		labelMap := make(map[string]string)
-		labelMap["topic"] = "topic1"
+		annotationlMap := make(map[string]string)
```
Comment (ewilde) [inline types/function_list_builder_test.go]: typo should be `annotationMap`

**Comment by ivanayov** [comment]
@ewilde Thanks for reviewing.
I've opened [#29](
https://github.com/openfaas/faas-swarm/issues/29) as this seems to be a bug.

**Comment by ivanayov** [comment]
Needs change in the README for using annotations, but this requires [faas-cli#498](https://github.com/openfaas/faas-cli/issues/498) first

**Comment by alexellis** [comment]
LGTM

**Pull request merged**